commons-math/src/main/java/org/apache/commons/math4/complex/Complex.java

```diff
@@ -0,0 +1,1308 @@
+
+        if (real == 0.0 && imaginary == 0.0) {
+            return createComplex(0.0, 0.0);
+        }
+
+        double t = FastMath.sqrt((FastMath.abs(real) + abs()) / 2.0);
+        if (real >= 0.0) {
+            return createComplex(t, imaginary / (2.0 * t));
+        } else {
+            return createComplex(FastMath.abs(imaginary) / (2.0 * t),
+                                 FastMath.copySign(1d, imaginary) * t);
+        }
+    }
+
+    /**
+     * Compute the
+     * <a href="http://mathworld.wolfram.com/SquareRoot.html" TARGET="_top">
+     * square root</a> of <code>1 - this<sup>2</sup></code> for this complex
+     * number.
+     * Computes the result directly as
+     * {@code sqrt(ONE.subtract(z.multiply(z)))}.
+     * <p>
+     * Returns {@link Complex#NaN} if either real or imaginary part of the
+     * input argument is {@code NaN}.
+     * </p>
+     * Infinite values in real or imaginary parts of the input may result in
+     * infinite or NaN values returned in parts of the result.
+     *
+     * @return the square root of <code>1 - this<sup>2</sup></code>.
+     * @since 1.2
+     */
+    public Complex sqrt1z() {
+        return createComplex(1.0, 0.0).subtract(this.multiply(this)).sqrt();
+    }
+
+    /**
+     * Compute the
+     * <a href="http://mathworld.wolfram.com/Tangent.html" TARGET="_top">
+     * tangent</a> of this complex number.
+     * Implements the formula:
+     * <pre>
+     *  <code>
+     *   tan(a + bi) = sin(2a)/(cos(2a)+cosh(2b)) + [sinh(2b)/(cos(2a)+cosh(2b))]i
+     *  </code>
+     * </pre>
+     * where the (real) functions on the right-hand side are
+     * {@link FastMath#sin}, {@link FastMath#cos}, {@link FastMath#cosh} and
+     * {@link FastMath#sinh}.
+     * <p>
+     * Returns {@link Complex#NaN} if either real or imaginary part of the
+     * input argument is {@code NaN}.
+     * </p>
+     * Infinite (or critical) values in real or imaginary parts of the input may
+     * result in infinite or NaN values returned in parts of the result.
+     * <pre>
+     *  Examples:
+     *  <code>
+     *   tan(a &plusmn; INFINITY i) = 0 &plusmn; i
+     *   tan(&plusmn;INFINITY + bi) = NaN + NaN i
+     *   tan(&plusmn;INFINITY &plusmn; INFINITY i) = NaN + NaN i
+     *   tan(&plusmn;&pi;/2 + 0 i) = &plusmn;INFINITY + NaN i
+     *  </code>
+     * </pre>
+     *
+     * @return the tangent of {@code this}.
+     * @since 1.2
+     */
+    public Complex tan() {
+        if (isNaN || Double.isInfinite(real)) {
+            return NaN;
+        }
+        if (imaginary > 20.0) {
+            return createComplex(0.0, 1.0);
+        }
+        if (imaginary < -20.0) {
+            return createComplex(0.0, -1.0);
+        }
+
+        double real2 = 2.0 * real;
+        double imaginary2 = 2.0 * imaginary;
+        double d = FastMath.cos(real2) + FastMath.cosh(imaginary2);
+
+        return createComplex(FastMath.sin(real2) / d,
+                             FastMath.sinh(imaginary2) / d);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 1092:1107, 1143:1157.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.